rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java

```diff
@@ -206,23 +201,25 @@ private boolean allowByMinerNotBanned(Peer sender, Message message) {
      * record event if message is repeated
      */
     private boolean allowByMessageUniqueness(Peer sender, Message message) {
+        if (message.getMessageType() != MessageType.BLOCK_MESSAGE && message.getMessageType() != MessageType.TRANSACTIONS) {
+            return true;
+        }
+
         Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
 
-        boolean contains = receivedMessages.contains(encodedMessage);
+        synchronized (receivedPeerMessagesMap) {
+            Set<NodeID> nodeIds = receivedPeerMessagesMap.getOrDefault(encodedMessage, new HashSet<>());
 
-        if (!contains) {
-            if (message.getMessageType() == MessageType.BLOCK_MESSAGE || message.getMessageType() == MessageType.TRANSACTIONS) {
-                if (this.receivedMessages.size() >= MAX_NUMBER_OF_MESSAGES_CACHED) {
-                    this.receivedMessages.clear();
-                }
-                this.receivedMessages.add(encodedMessage);
+            if (nodeIds.contains(sender.getPeerNodeID())) {
+                reportEventToPeerScoring(sender, EventType.REPEATED_MESSAGE, "Received repeated message on {}, not added to the queue");
+                return false;
             }
 
-        } else {
-            reportEventToPeerScoring(sender, EventType.REPEATED_MESSAGE, "Received repeated message on {}, not added to the queue");
+            nodeIds.add(sender.getPeerNodeID());
+            receivedPeerMessagesMap.put(encodedMessage, nodeIds);
         }
 
-        return !contains;
+        return true;
```

Comment:
```suggestion
        return noDuplicateDetected;
```

before your change this method was returning `!contains`, which means that false was being returned even if a duplicate message arrived from another peer, not necessary same peer. I'd try to preserve the previous logic of what was returned before.

Regarding the peer scoring punishment logic, it LGTM